# 1.0 `About Dataset`
Context
While many public datasets (on Kaggle and the like) provide Apple App Store data, there are not many counterpart datasets available for Google Play Store apps anywhere on the web. On digging deeper, I found out that iTunes App Store page deploys a nicely indexed appendix-like structure to allow for simple and easy web scraping. On the other hand, Google Play Store uses sophisticated modern-day techniques (like dynamic page load) using JQuery making scraping more challenging.

### 1.1 Content
Each app (row) has values for catergory, rating, size, and more.

### 1.2 Acknowledgements
This information is scraped from the Google Play Store. This app information would not be available without it.

### 1.3 Inspiration
The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market!

### 1.4 Dataset link:
> The Data Set was downloaded from Kaggle, from the following [link](https://www.kaggle.com/datasets/lava18/google-play-store-apps/)

# 2.0 `Import Libraries:`

In [2]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# 3.0 `Exploring Data:`

In [3]:
# load data
df = pd.read_csv('../../datasets/googleplaystore.csv')

In [4]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size     Installs  Type Price Content Rating  \
0      159   19M      10,000+  Free     0       Everyone   
1      967   14M     500,000+  Free     0       Everyone   
2    87510  8.7M   5,000,000+  Free     0       Everyone   
3   215644   25M  50,000,000+  Free     0           Teen   
4      967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [5]:
df.tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating  Reviews                Size     Installs  Type Price  \
10836     4.5       38                 53M       5,000+  Free     0   
10837     5.0        4                3.6M         100+  Free     0   
10838     NaN        3                9.5M       1,000+  Free     0   
10839     4.5      114  Varies with device       1,000+  Free     0   
10840     4.5   398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [6]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10841 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10839 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 1.1+ MB


In [8]:
df.describe()

Rating       Reviews
count  9367.000000  1.084100e+04
mean      4.191513  4.441119e+05
std       0.515735  2.927629e+06
min       1.000000  0.000000e+00
25%       4.000000  3.800000e+01
50%       4.300000  2.094000e+03
75%       4.500000  5.476800e+04
max       5.000000  7.815831e+07

In [9]:
df.isnull().sum()

App                  0
Category             1
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               1
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

In [10]:
df.isnull().sum()/len(df)*100

App                0.000000
Category           0.009224
Rating            13.596532
Reviews            0.000000
Size               0.000000
Installs           0.000000
Type               0.009224
Price              0.000000
Content Rating     0.000000
Genres             0.009224
Last Updated       0.000000
Current Ver        0.073794
Android Ver        0.018448
dtype: float64

## 3.11 `Observations:`

In [11]:
# showing all rows
pd.set_option('display.max_rows', None)

In [12]:
# showing unique values in each column with df.head()
df.columns.to_series().apply(lambda x: df[x].nunique())

App               9660
Category            33
Rating              39
Reviews           6001
Size               461
Installs            21
Type                 2
Price               92
Content Rating       6
Genres             119
Last Updated      1377
Current Ver       2831
Android Ver         33
dtype: int64

` Following are some anomalies in the data`
1. Rating has a maximum value of 19
2. Size has 'Varies with device',and  it should be numeric
3. Installs has '+' and ',', it should be numeric
4. Type has 'Free' and 'Paid'
5. Price has '$' and 'Everyone', it should be numeric
6. Content Rating has 'Adults only 18+' and 'Unrated'
7. Genres has ';'

# 4.0 `Exploring and Dealing with Columns:`

`What things to explore in every column of the dataset:`
1. Missing Values
2. DataType
3. Unique Values
4. Duplicates
5. Outliers
6. Distribution of data
7. Relationship between columns
8. Number of Apps in each Column(value_counts)

## 4.1 `App`:
### 4.1.1 `Exploring:`

In [13]:
df.App.isnull().sum()

0

In [14]:
df.App.nunique()

9660

In [15]:
# check for duplicates
df.duplicated().sum()

483

In [16]:
# DataType
df['App'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10841 entries, 0 to 10840
Series name: App
Non-Null Count  Dtype 
--------------  ----- 
10841 non-null  object
dtypes: object(1)
memory usage: 84.8+ KB


In [17]:
df['App'].value_counts().sum()

10841

In [18]:
df[df.duplicated(keep=False)].head()

App             Category  Rating  Reviews  \
164                   Ebook Reader  BOOKS_AND_REFERENCE     4.1    85842   
192  Docs To Go™ Free Office Suite             BUSINESS     4.1   217730   
193             Google My Business             BUSINESS     4.4    70991   
204                            Box             BUSINESS     4.2   159872   
213            ZOOM Cloud Meetings             BUSINESS     4.4    31614   

                   Size     Installs  Type Price Content Rating  \
164                 37M   5,000,000+  Free     0       Everyone   
192  Varies with device  50,000,000+  Free     0       Everyone   
193  Varies with device   5,000,000+  Free     0       Everyone   
204  Varies with device  10,000,000+  Free     0       Everyone   
213                 37M  10,000,000+  Free     0       Everyone   

                Genres   Last Updated         Current Ver         Android Ver  
164  Books & Reference  June 25, 2018               5.0.6          4.0 and up  
192           Business  April 2, 2018  Varies with device  Varies with device  
193           Business  July 24, 2018    2.19.0.204537701          4.4 and up  
204           Business  July 31, 2018  Varies with device  Varies with device  
213           Business  July 20, 2018      4.1.28165.0716          4.0 and up

In [19]:
df[df.duplicated(keep=False)].tail()

App      Category  Rating  \
8643                    Wunderlist: To-Do List & Tasks  PRODUCTIVITY     4.6   
8654   TickTick: To Do List with Reminder, Day Planner  PRODUCTIVITY     4.6   
8658                           ColorNote Notepad Notes  PRODUCTIVITY     4.6   
10049        Airway Ex - Intubate. Anesthetize. Train.       MEDICAL     4.3   
10768                                             AAFP       MEDICAL     3.8   

       Reviews                Size      Installs  Type Price Content Rating  \
8643    404610  Varies with device   10,000,000+  Free     0       Everyone   
8654     25370  Varies with device    1,000,000+  Free     0       Everyone   
8658   2401017  Varies with device  100,000,000+  Free     0       Everyone   
10049      123                 86M       10,000+  Free     0       Everyone   
10768       63                 24M       10,000+  Free     0       Everyone   

             Genres    Last Updated         Current Ver         Android Ver  
8643   Productivity   April 6, 2018  Varies with device  Varies with device  
8654   Productivity  August 6, 2018  Varies with device  Varies with device  
8658   Productivity   June 27, 2018  Varies with device  Varies with device  
10049       Medical    June 1, 2018              0.6.88          5.0 and up  
10768       Medical   June 22, 2018               2.3.1          5.0 and up

### 4.1.2 `Observations:`

My observations about app column are:
 1. There are 483 duplicates which are incorrect , it means that there are no duplicates in the App Column:
 2. There is no missing value
 2. There are 9660 unique values
 3. The data type is object


---

## 4.2 `Category`:
### 4.2.1 `Exploring:`

In [20]:
df.head(1)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   

   Reviews Size Installs  Type Price Content Rating        Genres  \
0      159  19M  10,000+  Free     0       Everyone  Art & Design   

      Last Updated Current Ver   Android Ver  
0  January 7, 2018       1.0.0  4.0.3 and up

In [21]:
df['Category'].value_counts()

Category
FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
Name: count, dtype: int64

In [22]:
df.Category.nunique()

33

In [23]:
# Missing Values
# Dealing With Missing Values Is Compulsory
df['Category'].isnull().sum()

1

In [24]:
df.Category.duplicated().sum()

10807

In [25]:
df[df.duplicated(subset='Category', keep=False)].head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size     Installs  Type Price Content Rating  \
0      159   19M      10,000+  Free     0       Everyone   
1      967   14M     500,000+  Free     0       Everyone   
2    87510  8.7M   5,000,000+  Free     0       Everyone   
3   215644   25M  50,000,000+  Free     0           Teen   
4      967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [26]:
df[df.duplicated(subset='Category', keep=False)].tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating  Reviews                Size     Installs  Type Price  \
10836     4.5       38                 53M       5,000+  Free     0   
10837     5.0        4                3.6M         100+  Free     0   
10838     NaN        3                9.5M       1,000+  Free     0   
10839     4.5      114  Varies with device       1,000+  Free     0   
10840     4.5   398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

### 4.2.2 `Observations:`

Following are my Observations about the Category Column:
1. Total categories in this dataset are 33
2. Total number of missing value is 1
3. There are no Duplicate rows, the duplicates it shows are not correct as many apps can have the same category, and there are no duplicate app in the dataset
4. Top Categories based on number of Apps are the Following:
- FAMILY     = 1972
- GAME        = 1144
- TOOLS        = 843
- MEDICAL      = 463
- BUSINESS     = 460

### 4.2.3 `Dealing With Column:`

#### 4.2.3.1 `Dealing With Missing Values:`

In [27]:
import warnings
# removing warnings
warnings.filterwarnings('ignore')

This column have only one missing value so we can drop it

In [28]:
# droping missing value row
df.dropna(subset=['Category'], inplace=True)

In [29]:
df['Category'].isnull().sum()

0

---

## 4.3 `Rating`:
### 4.3.1 `Exploring:`

In [30]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews Size  Installs  Type Price Content Rating  \
0      159  19M   10,000+  Free     0       Everyone   
1      967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [31]:
df.Rating.isnull().sum()/len(df)*100

13.597785977859777

In [32]:
df.Rating.isnull().sum()

1474

In [33]:
df['Rating'].nunique()

39

In [34]:
df['Rating'].value_counts()

Rating
4.4    1109
4.3    1076
4.5    1038
4.2     952
4.6     823
4.1     708
4.0     568
4.7     499
3.9     386
3.8     303
5.0     274
3.7     239
4.8     234
3.6     174
3.5     163
3.4     128
3.3     102
4.9      87
3.0      83
3.1      69
3.2      64
2.9      45
2.8      42
2.7      25
2.6      25
2.5      21
2.3      20
2.4      19
1.0      16
2.2      14
1.9      13
2.0      12
1.7       8
1.8       8
2.1       8
1.6       4
1.4       3
1.5       3
1.2       1
Name: count, dtype: int64

In [35]:
# number of apps with rating 4 and above
df[df['Rating'] >= 4].shape[0]

7368

In [36]:
df[(df['Rating'] < 4) & (df['Rating'] > 2)].shape[0]

1930

In [37]:
# number of apps with rating 2 and below
df[df['Rating']<=2].shape[0]

68

### 4.3.2 `Observations:`

I have done the following observations from the Rating Column:
1. Rating have 1474 Missing Values
2. Rating Column have 39 unique Values
3. Apps with rating 4 and above are `7368`
4. Apps with rating 2 to 4 are `1930`
5. Apps with rating 2 and below are `69`

### 4.3.3 `Dealing With Column:`

#### 4.3.3.1 `Dealing With Missing Values:`

As this Column is directly correlated with other columns so we have to deal with its missing values after some tests at the end of the notebook

---

## 4.4 `Reviews`:
### 4.4.1 `Exploring:`      

In [38]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews Size  Installs  Type Price Content Rating  \
0      159  19M   10,000+  Free     0       Everyone   
1      967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [39]:
df['Reviews'].info()

<class 'pandas.core.series.Series'>
Index: 10840 entries, 0 to 10840
Series name: Reviews
Non-Null Count  Dtype
--------------  -----
10840 non-null  int64
dtypes: int64(1)
memory usage: 169.4 KB


In [40]:
df['Reviews'].isnull().sum()

0

In [41]:
df['Reviews'].nunique()

6001

In [42]:
df[df['Reviews']>1000000].shape[0]

704

In [43]:
df[df['Reviews']>100000].shape[0]

2169

In [44]:
df[df['Reviews']>10000].shape[0]

4280

In [45]:
df[df['Reviews']<1000].shape[0]

4939

### 4.4.2 `Observations:`

I have done the following observations on Reviews column:

1. There are `no missing` values in the Reviews column.
2. The Reviews column is of the `int` data type.
3. There are` no duplicates `in the Reviews column.
4. There are `704` apps with more than 1,000,000 reviews.
5. There are `2169` apps with more than 100,000 reviews.
6. There are `4280` apps with more than 10,000 reviews.
7. There are `4940` apps with less than 1,000 reviews.

### 4.4.3 `Dealing With Column :`

There is nothing to deal with in this column but we have to feature engineering in this column later for better analysis.

---

## 4.5 `Size`:
### 4.5.1 `Exploring:`

In [46]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews Size  Installs  Type Price Content Rating  \
0      159  19M   10,000+  Free     0       Everyone   
1      967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [47]:
df['Size'].info()

<class 'pandas.core.series.Series'>
Index: 10840 entries, 0 to 10840
Series name: Size
Non-Null Count  Dtype 
--------------  ----- 
10840 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [48]:
df['Size'].isnull().sum()

0

In [49]:
df['Size'].unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

In [50]:
df['Size'].nunique()

461

In [51]:
df.Size.value_counts()

Size
Varies with device    1695
11M                    198
12M                    196
14M                    194
13M                    191
15M                    184
17M                    160
19M                    154
26M                    149
16M                    149
25M                    143
20M                    139
21M                    138
10M                    136
24M                    136
18M                    133
23M                    117
22M                    114
29M                    103
27M                     97
28M                     95
30M                     84
33M                     79
3.3M                    77
37M                     76
35M                     72
31M                     70
2.9M                    69
2.5M                    68
2.3M                    68
2.8M                    65
3.4M                    65
32M                     63
34M                     63
3.7M                    63
40M                     62
3.9M                   

The Value Count will be shown accurately when converted k to MB, and it is converted to numeric data type by removing M , and k

### 4.5.2 `Observations:`

`I have done the following observation on the Size column:`
1. The column has 461 unique values
2. The column has 0 missing value
3. The Column have an anomily that it should be numerical but it is object
4. It have two type of values, one in MB and other in KB, we need to convert it to single unit
5. Then we have to convert it to numeric by removing the 'k' and 'M' from the column
6. The column has 1 unique value 'Varies with device' which is not a valid value

### 4.5.3 `Dealing With Column:`

#### 4.5.3.1 `Converting kb to mbs:`

In [52]:
def convert_size(size):
    '''this function will convert the size from string to float, if the string contains kb then it will multiply the value by 1024 and if it contains M then it will multiply the value by 1024*1024'''
    if isinstance(size,str):
        if 'k' in size:
            return float(size.replace('k', ''))*1024
        if 'M' in size:
            return float(size.replace('M',''))*1024*1024
        elif size == 'Varies with device':
            return np.nan
    return size

In [53]:
df['Size'] = df['Size'].apply(convert_size)

In [54]:
df['Size'].info()

<class 'pandas.core.series.Series'>
Index: 10840 entries, 0 to 10840
Series name: Size
Non-Null Count  Dtype  
--------------  -----  
9145 non-null   float64
dtypes: float64(1)
memory usage: 169.4 KB


#### 4.5.3.2 `Renaming Column:`

In [55]:
# renaming the Size column to size_in_bytes
df.rename(columns={'Size':'Size_in_Bytes'}, inplace=True)

#### 4.5.3.3 `Adding Column size_in_mbs:`

In [56]:
# Making another Column of Size_in_Mbs
df['size_in_mbs'] = df['Size_in_Bytes']/1024/1024

#### 4.5.3.4 `Dealing With Missing Values:`

In [57]:
# first of all we have to check its corellation with other numeric columns
numeric_col = df[['Rating', 'Reviews', 'Size_in_Bytes', 'size_in_mbs']]

In [58]:
numeric_col.corr()

Rating   Reviews  Size_in_Bytes  size_in_mbs
Rating         1.000000  0.068141       0.083737     0.083737
Reviews        0.068141  1.000000       0.238214     0.238214
Size_in_Bytes  0.083737  0.238214       1.000000     1.000000
size_in_mbs    0.083737  0.238214       1.000000     1.000000

As I have not converted other object column to numeric column so I will deal with this missing Values at the end of notebook.

In [59]:
df['size_in_mbs'].isnull().sum()/len(df)*100

15.636531365313653

---

## 4.6 `Installs`:
### 4.6.1 `Exploring:`

In [60]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10,000+  Free     0       Everyone   
1      967     14680064.0  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [61]:
df.Installs.info()

<class 'pandas.core.series.Series'>
Index: 10840 entries, 0 to 10840
Series name: Installs
Non-Null Count  Dtype 
--------------  ----- 
10840 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [62]:
df.Installs.isnull().sum()

0

In [63]:
df.Installs.nunique()

21

In [64]:
df.Installs.unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

In [65]:
df.Installs.value_counts()

Installs
1,000,000+        1579
10,000,000+       1252
100,000+          1169
10,000+           1054
1,000+             907
5,000,000+         752
100+               719
500,000+           539
50,000+            479
5,000+             477
100,000,000+       409
10+                386
500+               330
50,000,000+        289
50+                205
5+                  82
500,000,000+        72
1+                  67
1,000,000,000+      58
0+                  14
0                    1
Name: count, dtype: int64

### 4.6.2 `Observations:`

`I have done the following observations from the Installs Column:`
1. The Installs column is of object data type.
2. The Installs column has no missing values.
3. The Installs column has `21` unique values.
4. The column have anomily of that it should be numeric but it is of object datatype, so we need to convert it to numeric.
5. It have to extra things first one is ',' and the other is '+'.
6. We need to remove these extra things and convert the column to numeric.
7. The column have 21 unique values, so we need to convert it to numeric and then we can use it for analysis.


### 4.6.3 `Dealing With Column:`

#### 4.6.3.1 `Removing ',' and '+' and Convert to numeric:`

In [66]:
def convert_numeric_installs(installs):
    '''this function will convert the installs from string to float, it will remove the , and + sign from the string and convert it to float'''
    if isinstance(installs, str):
        return float(installs.replace(',','').replace('+',''))
    return installs

In [67]:
df['Installs'] = df['Installs'].apply(convert_numeric_installs)

In [68]:
df['Installs'].info()

<class 'pandas.core.series.Series'>
Index: 10840 entries, 0 to 10840
Series name: Installs
Non-Null Count  Dtype  
--------------  -----  
10840 non-null  float64
dtypes: float64(1)
memory usage: 169.4 KB


In [69]:
df.Installs.head()

0       10000.0
1      500000.0
2     5000000.0
3    50000000.0
4      100000.0
Name: Installs, dtype: float64

---

## 4.7 `Type`:
### 4.7.1 `Exploring:`

In [70]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [71]:
df.Type.info()

<class 'pandas.core.series.Series'>
Index: 10840 entries, 0 to 10840
Series name: Type
Non-Null Count  Dtype 
--------------  ----- 
10839 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [72]:
df.Type.isnull().sum()

1

In [73]:
df.Type.nunique()

2

In [74]:
df.Type.unique()

array(['Free', 'Paid', nan], dtype=object)

In [75]:
df.Type.value_counts()

Type
Free    10039
Paid      800
Name: count, dtype: int64

### 4.7.2 `Observations:`

`I have done the following observations of the Type Column:`
1. Type column have `object` Data Type.
1. There are `2` unique values in the Type column which are `Free` and `Paid`.
2. The `Free` value has the `highest` number of counts.
3. The `Paid` value has the `lowest` number of counts.
4. There is `1` missing values in the Type column.

### 4.7.3 `Dealing With Column:`

#### 4.7.3.1 `Dealing With Missing value:`

In [76]:
# As there is only one missing value in the column so we can drop it
df.dropna(subset='Type', inplace=True)

In [77]:
df.Type.isnull().sum()

0

---

## 4.8 `Price`:
### 4.8.1 `Exploring:`

In [78]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [79]:
df.Price.info()

<class 'pandas.core.series.Series'>
Index: 10839 entries, 0 to 10840
Series name: Price
Non-Null Count  Dtype 
--------------  ----- 
10839 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [80]:
df.Price.isnull().sum()

0

In [81]:
df.Price.nunique()

92

In [82]:
df.Price.unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

In [83]:
df.Price.value_counts()

Price
0          10039
$0.99        148
$2.99        129
$1.99         73
$4.99         72
$3.99         63
$1.49         46
$5.99         30
$2.49         26
$9.99         21
$6.99         13
$399.99       12
$14.99        11
$4.49          9
$29.99         7
$24.99         7
$3.49          7
$7.99          7
$5.49          6
$19.99         6
$11.99         5
$6.49          5
$12.99         5
$8.99          5
$10.00         3
$16.99         3
$1.00          3
$2.00          3
$13.99         2
$8.49          2
$17.99         2
$1.70          2
$3.95          2
$79.99         2
$7.49          2
$9.00          2
$10.99         2
$39.99         2
$33.99         2
$1.96          1
$19.40         1
$4.80          1
$3.28          1
$4.59          1
$15.46         1
$3.04          1
$4.29          1
$2.60          1
$2.59          1
$3.90          1
$154.99        1
$4.60          1
$28.99         1
$2.95          1
$2.90          1
$1.97          1
$200.00        1
$89.99         1
$2.56   

### 4.8.2 `Observations:`

`I have done the following observations from the Price Column:`
1. The Price Column has `92` unique values.
2. The column have `no` missing values.
3. The column has `10041` apps with price value of `0`, which means that it is free.
4. The column is of `object` data type.
5. The column should be numeric for data analysis, so we have to remove `$` sign and convert it to `float` data type.

### 4.8.3 `Dealing With Column:`

#### 4.8.3.1 `Converting Column to Numeric DataType:`

In [84]:
# first of all to check that is there other currency included or not?

In [85]:
def convert_to_numeric_datatype(price):
    if isinstance(price, str):
        if '$' in price:
            return float(price.replace('$',''))
    return price

In [86]:
# now applying the function on Price column
df['Price'] = df['Price'].apply(convert_to_numeric_datatype)

---

## 4.9 `Content Rating`:
### 4.9.1 `Exploring:`

In [87]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [88]:
df['Content Rating'].info()

<class 'pandas.core.series.Series'>
Index: 10839 entries, 0 to 10840
Series name: Content Rating
Non-Null Count  Dtype 
--------------  ----- 
10839 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [89]:
df['Content Rating'].isnull().sum()

0

In [90]:
df['Content Rating'].nunique()

6

In [91]:
df['Content Rating'].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [92]:
df['Content Rating'].value_counts()

Content Rating
Everyone           8714
Teen               1208
Mature 17+          499
Everyone 10+        413
Adults only 18+       3
Unrated               2
Name: count, dtype: int64

### 4.9.2 `Observations:`

`I have done the following observations of the 'Content Rating' Column:`
1. The column has `no` missing values.
2. The column has `6` unique values.
3. These values are `'Everyone', 'Teen', 'Everyone 10+', 'Mature 17+', 'Adults only 18+', 'Unrated'`.
4. The value '`Everyone`' has the highest frequency of `8715`.
5. The value `'Adults only 18+`' has the lowest frequency of `3`.

---

## 4.10 `Genres`:
### 4.10.1 `Exploring:`

In [93]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [94]:
df.Genres.info()

<class 'pandas.core.series.Series'>
Index: 10839 entries, 0 to 10840
Series name: Genres
Non-Null Count  Dtype 
--------------  ----- 
10839 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [95]:
df.Genres.isnull().sum()

0

In [96]:
df.Genres.nunique()

119

In [97]:
df.Genres.unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

In [98]:
df.Genres.value_counts()

Genres
Tools                                    842
Entertainment                            623
Education                                549
Medical                                  463
Business                                 460
Productivity                             424
Sports                                   398
Personalization                          392
Communication                            387
Lifestyle                                381
Finance                                  366
Action                                   365
Health & Fitness                         341
Photography                              335
Social                                   295
News & Magazines                         283
Shopping                                 260
Travel & Local                           257
Dating                                   234
Books & Reference                        231
Arcade                                   220
Simulation                               200
Cas

### 4.10.2 `Observations:`

`I have done the following observations of the Genres Column`:
1. Genres Column have 1 missing Value.
2. Genres Column have 119 unique values.
3. The top 15 Genres with Values are :
   - Tools : 842
   - Entertainment : 623
   - Education : 549
   - Medical : 463
   - Business : 460
   - Productivity : 424
   - Sports : 398
   - Personalization : 392
   - Communication : 387
   - Lifestyle : 381
   - Finance : 366
   - Health & Fitness : 341
   - Photography : 335
   - Social : 295
   - News & Magazines : 283
4. The Column is Object Data Type.

### 4.10.3 `Dealing With Column:`

### 4.10.3.1 `Dealing Wih Missing Value:`

In [99]:
# as there is only one missing value so i will drop it, as genres of app cann't be changed
df.dropna(subset='Genres', inplace=True)

In [100]:
df['Genres'].isnull().sum()

0

---

## 4.11 `Last Updated`:
### 4.11.1 `Exploring:`

In [101]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [102]:
df['Last Updated'].info()

<class 'pandas.core.series.Series'>
Index: 10839 entries, 0 to 10840
Series name: Last Updated
Non-Null Count  Dtype 
--------------  ----- 
10839 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [103]:
df['Last Updated'].isnull().sum()

0

In [104]:
df['Last Updated'].nunique()

1377

In [105]:
df['Last Updated'].value_counts()

Last Updated
August 3, 2018        326
August 2, 2018        304
July 31, 2018         294
August 1, 2018        285
July 30, 2018         211
July 25, 2018         164
July 26, 2018         161
August 6, 2018        158
July 27, 2018         151
July 24, 2018         148
July 23, 2018         127
July 16, 2018         126
July 19, 2018         126
July 18, 2018         123
July 11, 2018         106
August 4, 2018        105
July 12, 2018         103
July 5, 2018           93
July 17, 2018          92
July 3, 2018           90
July 9, 2018           89
July 20, 2018          88
July 13, 2018          81
May 24, 2018           69
June 27, 2018          63
July 6, 2018           63
June 26, 2018          60
June 25, 2018          56
May 25, 2018           56
June 13, 2018          54
July 4, 2018           53
July 2, 2018           52
June 29, 2018          52
August 5, 2018         51
July 28, 2018          51
June 21, 2018          49
June 6, 2018           49
July 10, 2018          48

### 4.11.2 `Observations:`

`I have done the following Observations of the Last Updated Column`:
1. The Last Updated column is in `object` data type.
2. The Last Updated column has `no` missing values.
3. The Last Updated column has `1377` unique values.
4. Most of the Apps are lastly updated from `july 2018` to `august 2018`.
5. The Last Updated column has some Apps that were lastly updated between `2013 - 2015.`

---

## 4.12 `Current Version`:
### 4.12.1 `Exploring:`

In [106]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [107]:
df['Current Ver'].info()

<class 'pandas.core.series.Series'>
Index: 10839 entries, 0 to 10840
Series name: Current Ver
Non-Null Count  Dtype 
--------------  ----- 
10831 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [108]:
df['Current Ver'].isnull().sum()

8

In [109]:
df['Current Ver'].nunique()

2831

In [110]:
df['Current Ver'].value_counts()

Current Ver
Varies with device                                    1458
1.0                                                    809
1.1                                                    264
1.2                                                    178
2.0                                                    151
1.3                                                    145
1.0.0                                                  136
1.0.1                                                  119
1.4                                                     88
1.5                                                     81
1.0.2                                                   80
1.6                                                     65
1.0.3                                                   62
2.1                                                     61
3.0                                                     59
1.0.4                                                   58
1.7                                         

### 4.12.2 `Observations:`

` I have done the following observations from the Current Ver Column:`
1. The column has `2831` unique values.
2. The column has `8` missing values.
3. The Column is of `Object` DataType.
4. Most of the apps are in it `first version`.
5. There are very `few` apps that have been updated` more than twice`.
6. The Apps which are `updating frequently` are very `few` in numbers as compared to others.

### 4.12.3 `Dealing With Column:`

#### 4.12.3.1 `Dealing With Missing Values:`

In [111]:
# As there are only 8 missing values so we have to fill the missing values with mode
df['Current Ver'].fillna(df['Current Ver'].mode()[0], inplace =True)

In [112]:
# checking the missing values again
df['Current Ver'].isnull().sum()

0

---

## 4.13 `Andriod Version`:
### 4.13.1 `Exploring:`

In [113]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size_in_Bytes  Installs  Type Price Content Rating  \
0      159     19922944.0   10000.0  Free     0       Everyone   
1      967     14680064.0  500000.0  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  \
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up   

   size_in_mbs  
0         19.0  
1         14.0

In [114]:
df['Android Ver'].info()

<class 'pandas.core.series.Series'>
Index: 10839 entries, 0 to 10840
Series name: Android Ver
Non-Null Count  Dtype 
--------------  ----- 
10837 non-null  object
dtypes: object(1)
memory usage: 169.4+ KB


In [115]:
df['Android Ver'].isnull().sum()

2

In [116]:
df['Android Ver'].nunique()

33

In [117]:
df['Android Ver'].unique()

array(['4.0.3 and up', '4.2 and up', '4.4 and up', '2.3 and up',
       '3.0 and up', '4.1 and up', '4.0 and up', '2.3.3 and up',
       'Varies with device', '2.2 and up', '5.0 and up', '6.0 and up',
       '1.6 and up', '1.5 and up', '2.1 and up', '7.0 and up',
       '5.1 and up', '4.3 and up', '4.0.3 - 7.1.1', '2.0 and up',
       '3.2 and up', '4.4W and up', '7.1 and up', '7.0 - 7.1.1',
       '8.0 and up', '5.0 - 8.0', '3.1 and up', '2.0.1 and up',
       '4.1 - 7.1.1', nan, '5.0 - 6.0', '1.0 and up', '2.2 - 7.1.1',
       '5.0 - 7.1.1'], dtype=object)

In [118]:
df['Android Ver'].value_counts()

Android Ver
4.1 and up            2451
4.0.3 and up          1501
4.0 and up            1375
Varies with device    1361
4.4 and up             980
2.3 and up             652
5.0 and up             601
4.2 and up             394
2.3.3 and up           281
2.2 and up             244
4.3 and up             243
3.0 and up             241
2.1 and up             134
1.6 and up             116
6.0 and up              60
7.0 and up              42
3.2 and up              36
2.0 and up              32
5.1 and up              24
1.5 and up              20
4.4W and up             12
3.1 and up              10
2.0.1 and up             7
8.0 and up               6
7.1 and up               3
4.0.3 - 7.1.1            2
5.0 - 8.0                2
1.0 and up               2
7.0 - 7.1.1              1
4.1 - 7.1.1              1
5.0 - 6.0                1
2.2 - 7.1.1              1
5.0 - 7.1.1              1
Name: count, dtype: int64

### 4.13.2 `Observations:`

` I have done the Following observations from the 'Andriod Version' Column:`
1. The column is of `object` data type.
2. The column have `2` missing Values.
3. The Column have `33` unique values.
4. The Lowest Andriod Version is `1` and the highest column is `8`.
5. The Most Common Andriod Version is `4.1` and `4.0.3`, which have the greatest number of apps in the dataset.
6. Andriod version `8.0` and up have just `6` apps.
7. Column have Unique value '`Varies with device`', which does not specifies any andriod version, so this is the anomily in the dataset.

### 4.13.3 `Dealing With Column:`

#### 4.13.3.1 `Dealing With Missing Values:`

In [119]:
df['Android Ver'].fillna(df['Android Ver'].mode()[0], inplace=True)

# 5.0 `Dealing With Remaining Missing Values`:

## 5.1 `Checking Missing Values`:

In [120]:
df.isnull().sum()

App                  0
Category             0
Rating            1473
Reviews              0
Size_in_Bytes     1694
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          0
Android Ver          0
size_in_mbs       1694
dtype: int64

## 5.2 `Checking Correlation of The Numerical Columns in the Dataset`:

In [121]:
# I have to make a datafrome of numeric column from this dataset
numeric_col = df.select_dtypes(include=['int64', 'float64'])

In [122]:
numeric_col.head()

Rating  Reviews  Size_in_Bytes    Installs  size_in_mbs
0     4.1      159     19922944.0     10000.0         19.0
1     3.9      967     14680064.0    500000.0         14.0
2     4.7    87510      9122611.2   5000000.0          8.7
3     4.5   215644     26214400.0  50000000.0         25.0
4     4.3      967      2936012.8    100000.0          2.8

In [123]:
numeric_col.corr()

Rating   Reviews  Size_in_Bytes  Installs  size_in_mbs
Rating         1.000000  0.068141       0.083737  0.051355     0.083737
Reviews        0.068141  1.000000       0.238214  0.643121     0.238214
Size_in_Bytes  0.083737  0.238214       1.000000  0.164787     1.000000
Installs       0.051355  0.643121       0.164787  1.000000     0.164787
size_in_mbs    0.083737  0.238214       1.000000  0.164787     1.000000

## 5.3 `Performing Tests`:

### 5.3.1 `Pearson's Test`:

In [124]:
from scipy import stats
from scipy.stats import pearsonr


In [125]:
size.info()

NameError: name 'size' is not defined

In [ ]:
installs = df['Installs']

In [ ]:
installs.info()

<class 'pandas.core.series.Series'>
Index: 9145 entries, 0 to 9144
Series name: Installs
Non-Null Count  Dtype  
--------------  -----  
9145 non-null   float64
dtypes: float64(1)
memory usage: 142.9 KB


In [ ]:
# 10839 in installs and 9145 in size, so we have to remove the remaining values from the installs dataframe
installs = installs[:9145]

In [ ]:
# making the length of size and installs equal
# minimize the lenth of installs column same to the size column
installs = df['Installs']

In [ ]:
r, p = pearsonr(size, installs)  # or pearsonr(size_mbs, reviews)


In [ ]:
print("Pearson's r: ", r)
print("p-value: ", p)

Pearson's r:  0.002674706972317305
p-value:  0.7981466905038033


Based on the output:

- Pearson's r: 0.002674706972317305:
    - This is the correlation coefficient, which measures the strength and direction of the linear relationship between the two variables.
    - The value is very close to 0, indicating a extremely weak correlation between the variables.
- p-value: 0.7981466905038033:
    - This is the probability of observing a correlation coefficient at least as extreme as the one you observed, assuming that there is no true correlation between the variables.
    - The p-value is very high (close to 1), indicating that the observed correlation is likely due to chance (i.e., not statistically significant).

In summary, the results suggest that there is no significant linear correlation between the variables you tested (likely size_in_mbs and Installs or Reviews).


## 5.4 `Dealing Missing Values`:

### 5.4.1 `Review Column`:

In [ ]:
df['Reviews'].fillna(df['Reviews'].median(), inplace =True)

In [ ]:
df.Reviews.isnull().sum()

0

### 5.4.2 `Size Columns`:

In [ ]:
# we will replace the missing values with the mean of the column
df['size_in_mbs'].fillna(df.size_in_mbs.mean(), inplace= True)


In [ ]:
df.size_in_mbs.isnull().sum()

0

In [ ]:
df['Size_in_Bytes'].fillna(df.Size_in_Bytes.mean(), inplace= True)


In [ ]:
df.Size_in_Bytes.isnull().sum()

0

In [127]:
# now to save the cleaned data
df.to_csv('../../datasets/cleaned_data/cleaned_googleplaystore.csv', index=False)